tests/adjoint_2d_cylindrical/forward.py

```diff
@@ -0,0 +1,236 @@
+    averager = LayerAveraging(
+        mesh, np.linspace(geo_constants["rmin"], geo_constants["rmax"], geo_constants["nlayers"] * 2), quad_degree=6
+    )
+    averager.extrapolate_layer_average(Taverage, averager.get_layer_average(T))
+
+    checkpoint_file = CheckpointFile("Checkpoint_State.h5", "w")
+    checkpoint_file.save_mesh(mesh)
+    checkpoint_file.save_function(Taverage, name="Average_Temperature", idx=0)
+    checkpoint_file.save_function(T, name="Temperature", idx=0)
+
+    Z_nullspace = create_stokes_nullspace(Z, closed=True, rotational=True)
+    Z_near_nullspace = create_stokes_nullspace(
+        Z, closed=False, rotational=True, translations=[0, 1]
+    )
+
+    # Free-slip velocity boundary condition on all sides
+    stokes_bcs = {
+        "bottom": {"un": 0},
+        "top": {"un": 0},
+    }
+    temp_bcs = {
+        "bottom": {"T": 1.0},
+        "top": {"T": 0.0},
+    }
+
+    energy_solver = EnergySolver(
+        T, u, approximation, delta_t, ImplicitMidpoint, bcs=temp_bcs
+    )
+
+    stokes_solver = StokesSolver(
+        z,
+        T,
+        approximation,
+        bcs=stokes_bcs,
+        solver_parameters="direct",
+        nullspace=Z_nullspace,
+        transpose_nullspace=Z_nullspace,
+        near_nullspace=Z_near_nullspace,
+    )
+
+    if visualise:
+        # Create output file and select output_frequency
+        output_file = VTKFile("vtu-files/output.pvd")
+        dump_period = 10
+
+    # Now perform the time loop:
+    for timestep in range(0, max_timesteps):
+        stokes_solver.solve()
+        energy_solver.solve()
+
+        # Storing velocity to be used in the objective F
+        checkpoint_file.save_function(z.subfunctions[0], name="Velocity", idx=timestep)
+
+        if visualise:
+            if (timestep % dump_period == 0 or timestep == max_timesteps - 1):
+                mu_function.interpolate(mu)
+                output_file.write(*z.subfunctions, T, mu_function)
+
+    # Save the reference final temperature
+    checkpoint_file.save_function(T, name="Temperature", idx=max_timesteps - 1)
+    checkpoint_file.close()
+
+
+def get_geometry_parameters():
+    """
+    Returns the geometry parameters for the 2D cylindrical adjoint test case.
+    """
+    rmin, rmax, nlayers = 1.22, 2.22, 128
+    rmax_earth = 6370  # Radius of Earth [km]
+    rmin_earth = rmax_earth - 2900  # Radius of CMB [km]
+    r_410_earth = rmax_earth - 410  # 410 radius [km]
+    r_660_earth = rmax_earth - 660  # 660 raidus [km]
+    r_410 = rmax - (rmax_earth - r_410_earth) / (rmax_earth - rmin_earth)
+    r_660 = rmax - (rmax_earth - r_660_earth) / (rmax_earth - rmin_earth)
+
+    return {
+        "rmin": rmin,
+        "rmax": rmax,
+        "nlayers": nlayers,
+        "rmax_earth": rmax_earth,
+        "rmin_earth": rmin_earth,
+        "r_410": r_410,
+        "r_660": r_660,
+        "r_410_earth": r_410_earth,
+        "r_660_earth": r_660_earth,
+    }
```

Comment:
```python
    mantle_thickness = 2900
    r_410 = rmax - 410 / mantle_thickness
    r_660 = rmax - 660 / mantle_thickness

    return {
        "rmin": rmin,
        "rmax": rmax,
        "nlayers": nlayers,
        "r_410": r_410,
        "r_660": r_660,
    }
```

Reply:
Up to you. :)
My intention was to improve readability, but if you think the additional parameters will be needed in the future, then fine to keep them in.